File: server/task/task.service.js
import { v4 } from 'uuid';

export default class TaskService {

    constructor(repo) {
        this.repo = repo;
    }

    addTask(projectId, description) {
        const id = v4();
        const task = { id, projectId, description, createdAt: new Date(), finishedAt: null };
        return this.repo.create(task)
            .then(_res => ({ status: 'OK', taskId: id, message: 'Task was created.' }))
    }

    updateTask(projectId, task) {
        return this.repo.isFinished(projectId, task.taskId)
            .then(wasFinished => {
                if (wasFinished) {
                    return { status:'OK', message: 'This task has already been finished.' };
                }
                return this.repo.update({ id: task.taskId, projectId, description: task.description })
                    .then(_res => ({ status: 'OK', message: 'Task was updated.' }));
            });
    }

    finish(projectId, taskId) {
        return this.repo.isFinished(projectId, taskId)
            .then(wasFinished => {
                if (wasFinished) {
                    return { status:'OK', message: 'This task has already been finished.' };
                }
                return this.repo.finish(projectId, taskId)
                    .then(_res => ({ status: 'OK', message: 'Task was finished.' }));
            })
    }

    remove(projectId, taskId) {
        return this.repo.isFinished(projectId, taskId)
            .then(wasFinished => {
                if (wasFinished) {
                    return { status:'FAIL', message: 'This task cannot be removed.' };
                }
                return this.repo.remove(projectId, taskId);
            })
    }

    getTasks(projectId) {
        return this.repo.findtasks(projectId)
            .then(list => list.map(({ id, description, finishedAt }) => ({ id, description, finishedAt })));
    }

}

File: pkg/domain/task/task_service.go
package task

import (
	"errors"
	"time"

	"github.com/google/uuid"
	m "github.com/keuller/multi-task-manager/pkg/domain/model"
)

type TaskService struct {
	repo TaskRepository
}

func NewTaskService(repo TaskRepository) TaskService {
	return TaskService{repo}
}

func (s TaskService) AddTask(project string, data m.CreateTaskRequest) string {
	newId := uuid.Must(uuid.NewRandom())

	task := m.Task{
		ID:          newId.String(),
		Description: data.Description,
		ProjectID:   project,
		CreatedAt:   time.Now(),
	}

	s.repo.Create(project, task)
	return newId.String()
}

func (s TaskService) UpdateTask(project, task, desc string) error {
	if err := s.repo.Update(project, task, desc); err != nil {
		return err
	}
	return nil
}

func (s TaskService) FinishTask(project, task string) error {
	if err := s.repo.Finish(project, task); err != nil {
		return err
	}
	return nil
}

func (s TaskService) RemoveTask(project, task string) error {
	if !s.repo.IsFinished(project, task) {
		s.repo.Remove(project, task)
		return nil
	}
	return errors.New("task is finished and cannot be removed")
}

func (s TaskService) GetByProject(project string) []m.TaskRecordResponse {
	tasks, exists := s.repo.data[project]
	if !exists {
		return nil
	}

	result := make([]m.TaskRecordResponse, 0)
	for _, t := range tasks {
		result = append(result, m.TaskRecordResponse{
			ID:          t.ID,
			Description: t.Description,
			CreatedAt:   t.CreatedAt,
			FinishedAt:  t.FinishedAt,
		})
	}

	return result
}
